onadata/apps/logger/migrations/0030_backfill_lost_monthly_counters.py

```diff
@@ -15,6 +15,20 @@ def populate_missing_monthly_counters(apps, schema_editor):
     if not DailyXFormSubmissionCounter.objects.all().exists():
         return
 
+    # Associate each daily counter with user=None with a user based on its xform
+    for counter in DailyXFormSubmissionCounter.objects.filter(user=None).iterator:
+        if counter.xform and counter.xform.user:
+            has_duplicate = DailyXFormSubmissionCounter.objects.filter(
+                date=counter.date, xform=counter.xform
+            ).exclude(user=None).exists()
+            # don't add a user to duplicate counters, so they get deleted in the next step
+            if not has_duplicate:
+                counter.user = counter.xform.user
+                counter.save()
+
+    # Delete daily counters without a user to avoid creating invalid monthly counters
+    DailyXFormSubmissionCounter.objects.filter(user=None).delete()
+
```

Comment:
I'm wondering if those lines should be present in the `0031` too for users who have already run `0030` on an earlier date. 

Example: 
- Upgrade to `2.023.37g` (which contains `0030`) on Oct 5. 
- Upgrade to `2.023.37h` (which contains `0031`) on Oct 13.

`0030` won't be reapplied and `null` could have been inserted between both patch releases. 

What about adding adding a `RunPython` in `0031` which would block the migration if `null` is detected in `user_id` field? We could provided instruction to rollback to `0029` and upgrade to `0031`. Otherwise, it would apply without any message.

Reply:
That's a very good point. I'm planning on refactoring `0030` to delete the daily counters and then repopulate the monthly counters, as well as creating a new `0031` migration that just deletes daily counters. The previous `0031` migration is now `0032`.

This covers the edge case where a user upgraded to `2.023.37g` on the same date as they upgrade to `2.023.37h`. (Also, it means one less manual step needed to deploy.)